### PR TITLE
SONAR-8429 add organization parameter to api/ce/submit

### PR DESCRIPTION
**Pull request by sns-seb**

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/computation/queue/ReportSubmitter.java [hunk 1]
```diff
@@ -62,17 +67,36 @@ public ReportSubmitter(CeQueue queue, UserSession userSession,
...
     }
   }
 
-  private ComponentDto createProject(DbSession dbSession, String projectKey, @Nullable String projectBranch, @Nullable String projectName) {
+  private OrganizationDto getOrganizationDtoOrFail(DbSession dbSession, String organizationKey) {
+    return dbClient.organizationDao().selectByKey(dbSession, organizationKey)
+      .orElseThrow(() -> new NotFoundException(format("Organization with key '%s' does not exist", organizationKey)));
+  }
+
+  private void ensureOrganizationIsConsistent(Optional<ComponentDto> project, OrganizationDto organizationDto) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/queue/ReportSubmitter.java]: MAJOR Make "ensureOrganizationIsConsistent" a "static" method. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2325)

File: server/sonar-server/src/main/java/org/sonar/server/ce/ws/SubmitAction.java [hunk 2]
```diff
@@ -79,13 +89,16 @@ public void define(WebService.NewController controller) {
 
   @Override
   public void handle(Request wsRequest, Response wsResponse) throws Exception {
+    String organizationKey = wsRequest.getParam(PARAM_ORGANIZATION_KEY)
+      .emptyAsNull()
+      .or(() -> defaultOrganizationProvider.get().getKey());
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/ce/ws/SubmitAction.java]: MINOR Replace this lambda with a method reference. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1612)

File: server/sonar-server/src/main/java/org/sonar/ce/queue/CeTask.java [hunk 3]
```diff
@@ -102,13 +111,24 @@ public int hashCode() {
...
     private String componentKey;
     private String componentName;
     private String submitterLogin;
 
+    public Builder setOrganizationUuid(String organizationUuid) {
+      this.organizationUuid = organizationUuid;
+      return this;
+    }
+
+    // FIXME remove this method when organization support is added to the Compute Engine queue
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/ce/queue/CeTask.java]: MAJOR Take the required action to fix the issue indicated by this comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1134)

File: server/sonar-server/src/main/java/org/sonar/ce/queue/CeQueueImpl.java [hunk 4]
```diff
@@ -213,10 +219,15 @@ public CeTask apply(@Nonnull CeQueueDto dto) {
         builder.setComponentUuid(componentUuid);
         ComponentDto component = componentDtoByUuid.get(componentUuid);
         if (component != null) {
+          builder.setOrganizationUuid(component.getOrganizationUuid());
           builder.setComponentKey(component.getKey());
           builder.setComponentName(component.name());
         }
       }
+      // fixme this should be set from the CeQueueDto
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/ce/queue/CeQueueImpl.java]: MAJOR Take the required action to fix the issue indicated by this comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1134)

File: server/sonar-db-migration/src/main/java/org/sonar/server/platform/db/migration/version/v63/DefaultOrganizationUuid.java [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.server.platform.db.migration.version.v63;
+
+import java.sql.SQLException;
+import org.sonar.server.platform.db.migration.step.DataChange;
+
+public interface DefaultOrganizationUuid {
```
Comment (SonarTech) [inline server/sonar-db-migration/src/main/java/org/sonar/server/platform/db/migration/version/v63/DefaultOrganizationUuid.java]: CRITICAL Annotate the "DefaultOrganizationUuid" interface with the @FunctionalInterface annotation [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1609)

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues
* CRITICAL 1 critical
* MAJOR 3 major
* MINOR 1 minor

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* MAJOR 1 major
* MINOR 1 minor

Watch the comments in this conversation to review them.

File: server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/LoadReportAnalysisMetadataHolderStep.java [hunk 6]
```diff
@@ -88,6 +103,51 @@ private void checkProjectKeyConsistency(ScannerReport.Metadata reportMetadata) {
...
+
+  private static boolean reportBelongsToDefaultOrganization(@Nullable String organizationKey) {
+    return organizationKey == null || organizationKey.isEmpty();
+  }
+
+  private Organization toOrganization(String organizationUuid) {
+    try (DbSession dbSession = dbClient.openSession(false)) {
+      Optional<OrganizationDto> organizationDto = dbClient.organizationDao().selectByUuid(dbSession, organizationUuid);
+      checkState(organizationDto.isPresent(), "Organization with uuid '{}' can't be found", organizationUuid);
+      return toOrganization(organizationDto.get());
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/LoadReportAnalysisMetadataHolderStep.java]: MAJOR Call "organizationDto.isPresent()" before accessing the value. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3655)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* MAJOR 1 major

Watch the comments in this conversation to review them.

File: server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/LoadReportAnalysisMetadataHolderStep.java [hunk 7]
```diff
@@ -88,6 +103,43 @@ private void checkProjectKeyConsistency(ScannerReport.Metadata reportMetadata) {
...
+
+  private static boolean reportBelongsToDefaultOrganization(@Nullable String organizationKey) {
+    return organizationKey == null || organizationKey.isEmpty();
+  }
+
+  private Organization toOrganization(String organizationUuid) {
+    try (DbSession dbSession = dbClient.openSession(false)) {
+      Optional<OrganizationDto> organizationDto = dbClient.organizationDao().selectByUuid(dbSession, organizationUuid);
+      checkState(organizationDto.isPresent(), "Organization with uuid '{}' can't be found", organizationUuid);
+      return Organization.from(organizationDto.get());
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/LoadReportAnalysisMetadataHolderStep.java]: MAJOR Call "organizationDto.isPresent()" before accessing the value. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3655)

File: sonar-scanner-protocol/src/main/protobuf/scanner_report.proto [hunk 8]
```diff
@@ -31,11 +31,12 @@ option optimize_for = SPEED;
 message Metadata {
   int64 analysis_date = 1;
   // TODO should we keep this project_key here or not ? Because it's a duplication of Component.key
```
Comment (simonbrandhof) [inline sonar-scanner-protocol/src/main/protobuf/scanner_report.proto]: out of scope of PR, but I think this TODO can be dropped.

File: server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/LoadReportAnalysisMetadataHolderStep.java [hunk 9]
```diff
@@ -88,6 +103,43 @@ private void checkProjectKeyConsistency(ScannerReport.Metadata reportMetadata) {
     }
   }
 
+  private void checkOrganizationKeyConsistency(ScannerReport.Metadata reportMetadata, Organization organization) {
+    String organizationKey = reportMetadata.getOrganizationKey();
+    String resolveReportOrganizationKey = resolveReportOrganizationKey(organizationKey);
+    if (!resolveReportOrganizationKey.equals(organization.getKey())) {
+      if (reportBelongsToDefaultOrganization(organizationKey)) {
+        throw MessageException.of(format(
+          "Report does not specify an OrganizationKey but it has been submitted to another organization (%s) than the default one (%s)",
```
Comment (simonbrandhof) [inline server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/LoadReportAnalysisMetadataHolderStep.java]: Because message targets end-users, I think "organizationKey" can be simply replaced by "organization" in the two messages (lines 112 and 117)

File: sonar-db/src/main/resources/org/sonar/db/component/ComponentMapper.xml [hunk 10]
```diff
@@ -65,12 +65,12 @@
...
   <select id="selectByProjectUuid" parameterType="string" resultType="Component">
-    SELECT
+    select
     <include refid="componentColumns"/>
-    FROM projects p
-    <where>
-      p.project_uuid=#{projectUuid}
-    </where>
+    from projects root
+    inner join projects p on p.project_uuid=root.uuid and p.organization_uuid=root.organization_uuid
```
Comment (simonbrandhof) [inline sonar-db/src/main/resources/org/sonar/db/component/ComponentMapper.xml]: As discussed SQL inner joins should be dropped. I suppose it will be done in another PR ?

File: server/sonar-server/src/main/java/org/sonar/server/ce/ws/SubmitAction.java [hunk 11]
```diff
@@ -27,20 +27,24 @@
...
 import org.sonar.server.ws.WsUtils;
 import org.sonarqube.ws.WsCe;
 
 public class SubmitAction implements CeWsAction {
 
-  public static final String PARAM_PROJECT_KEY = "projectKey";
-  public static final String PARAM_PROJECT_BRANCH = "projectBranch";
-  public static final String PARAM_PROJECT_NAME = "projectName";
-  public static final String PARAM_REPORT_DATA = "report";
+  private static final String PARAM_ORGANIZATION_KEY = "organization";
```
Comment (henryju) [inline server/sonar-server/src/main/java/org/sonar/server/ce/ws/SubmitAction.java]: Naming is not consistent. projectOrganization or projectOrg ?

Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/ce/ws/SubmitAction.java]: going this way, we will have to name the parameter qpOrganization on the QP webservices, qgOrganization on the QG webservices, ...
I don't think we want to take this road and Lambda already indicated that for consistency of the Web API as a whole, a given information should have the same parameter name accross all WS

Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/ce/ws/SubmitAction.java]: ping @julienlancelot for confirmation

Comment (simonbrandhof) [inline server/sonar-server/src/main/java/org/sonar/server/ce/ws/SubmitAction.java]: I think we should mix scanner configuration and web services.
IMO it makes sense that scanner accepts param "sonar.projectOrganization" whereas web services have param "organization".

File: server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/LoadReportAnalysisMetadataHolderStep.java (hunk 9 above)
Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/LoadReportAnalysisMetadataHolderStep.java]: Message does target the end user but the name of the field in the report is organizationKey. It will make it easier to this user if the name of the field with a wrong value is accurately reported, don't you think?

File: sonar-db/src/main/resources/org/sonar/db/component/ComponentMapper.xml (hunk 10 above)
Comment (sns-seb) [inline sonar-db/src/main/resources/org/sonar/db/component/ComponentMapper.xml]: yes it is

File: server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/LoadReportAnalysisMetadataHolderStep.java (hunk 9 above)
Comment (simonbrandhof) [inline server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/LoadReportAnalysisMetadataHolderStep.java]: In fact this error can only be a bug in scanner. So indeed message can be technically exact. It could also be a ISE, but that's minor.

File: server/sonar-server/src/main/java/org/sonar/server/ce/ws/SubmitAction.java (hunk 11 above)
Comment (henryju) [inline server/sonar-server/src/main/java/org/sonar/server/ce/ws/SubmitAction.java]: This has nothing to do with scanner. Other parameters of this WS are "projectKey", "projectBranch", "projectName", ... But this is minor :)

Comment (simonbrandhof) [inline server/sonar-server/src/main/java/org/sonar/server/ce/ws/SubmitAction.java]: arggh, you're right for this point !

Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/ce/ws/SubmitAction.java]: I already replied to the "WS naming consistency" comment

> I don't think we want to take this road and Lambda already indicated that for consistency of the Web API as a whole, a given information should have the same parameter name accross all WS

For this reason, I think the parameter should be named "organization"
